code/game/objects/items/pinpointer.dm

```diff
@@ -7,137 +11,114 @@
-			to_chat(usr, "<span class='notice'>You deactivate the pinpointer</span>")
+	var/mode = PINPOINTER_MODE_DISK_GENERATOR
 
 
-	proc/worklocation()
-		if(!active)
-			return
-		if(!location)
-			icon_state = "pinonnull"
-			return
-		setDir(get_dir(src,location))
-		switch(get_dist(src,location))
-			if(0)
-				icon_state = "pinondirect"
-			if(1 to 8)
-				icon_state = "pinonclose"
-			if(9 to 16)
-				icon_state = "pinonmedium"
-			if(16 to INFINITY)
-				icon_state = "pinonfar"
-		spawn(5) .()
-
-
-	proc/workobj()
-		if(!active)
-			return
-		if(!target)
-			icon_state = "pinonnull"
-			return
-		setDir(get_dir(src,target))
-		switch(get_dist(src,target))
-			if(0)
-				icon_state = "pinondirect"
-			if(1 to 8)
-				icon_state = "pinonclose"
-			if(9 to 16)
-				icon_state = "pinonmedium"
-			if(16 to INFINITY)
-				icon_state = "pinonfar"
-		spawn(5) .()
+/obj/item/pinpointer/advpinpointer/Initialize()
+	. = ..()
+	switch(mode)
+		if(PINPOINTER_MODE_DISK)
+			tracked_list = GLOB.nuke_disk_list
+		if(PINPOINTER_MODE_DISK_GENERATOR)
+			tracked_list = GLOB.nuke_disk_generators
+
+
+/obj/item/pinpointer/advpinpointer/examine(mob/user)
+	. = ..()
+	if(mode == PINPOINTER_MODE_DISK)
+		for(var/i in tracked_list)
+			var/obj/machinery/nuclearbomb/bomb = i
+			if(!bomb.timer_enabled)
+				continue
+			to_chat(user, "Extreme danger.  Arming signal detected.   Time remaining: [bomb.timeleft]")
+
 
 /obj/item/pinpointer/advpinpointer/verb/toggle_mode()
 	set category = "Object"
 	set name = "Toggle Pinpointer Mode"
 	set src in view(1)
 
-	active = 0
-	icon_state = "pinoff"
-	target=null
-	location = null
-
-	switch(alert("Please select the mode you want to put the pinpointer in.", "Pinpointer Mode Select", "Location", "Disk Recovery", "Other Signature"))
-		if("Location")
-			mode = 1
+	if(active)
+		deactivate()
 
-			var/locationx = input(usr, "Please input the x coordinate to search for.", "Location?" , "") as num
-			if(!locationx || !(usr in view(1,src)))
-				return
-			var/locationy = input(usr, "Please input the y coordinate to search for.", "Location?" , "") as num
-			if(!locationy || !(usr in view(1,src)))
-				return
-
-			var/turf/Z = get_turf(src)
-
-			location = locate(locationx,locationy,Z.z)
+	switch(alert("Please select the mode you want to put the pinpointer in.", "Pinpointer Mode Select", "Disk Recovery", "Disk Generator Tracking"))
+		if("Disk Recovery")
+			do_toggle_mode(PINPOINTER_MODE_DISK)
+			activate(usr)
+			return
+		if("Disk Generator Tracking")
+			do_toggle_mode(PINPOINTER_MODE_DISK_GENERATOR)
+			activate(usr)
+			return
 
-			to_chat(usr, "You set the pinpointer to locate [locationx],[locationy]")
 
+/obj/item/pinpointer/advpinpointer/proc/do_toggle_mode(new_mode)
+	switch(new_mode)
+		if(PINPOINTER_MODE_DISK)
+			tracked_list = GLOB.nuke_disk_list
+		if(PINPOINTER_MODE_DISK_GENERATOR)
+			tracked_list = GLOB.nuke_disk_generators
+		else
+			CRASH("do_toggle_mode called with an invalid new_mode")
```

Comment:
Will this ever happen? Psy didn't have this in his PR.

Reply:
Debug check for the future. It's within a `switch`, so it's free.

Reply:
Evil, but acceptable.